extension/fts/src/function/fts.cpp

```diff
@@ -0,0 +1,326 @@
+          avgDocLen{other.avgDocLen} {}
+
+    bool hasNodeInput() const override { return true; }
+    std::shared_ptr<binder::Expression> getNodeInput() const override { return terms; }
+
+    std::unique_ptr<GDSBindData> copy() const override {
+        return std::make_unique<FTSGDSBindData>(*this);
+    }
+};
+
+struct ScoreData {
+    uint64_t df;
+    uint64_t tf;
+
+    ScoreData(uint64_t df, uint64_t tf) : df{df}, tf{tf} {}
+};
+
+struct ScoreInfo {
+    nodeID_t termID;
+    std::vector<ScoreData> scoreData;
+
+    explicit ScoreInfo(nodeID_t termID) : termID{std::move(termID)} {}
+
+    void addEdge(uint64_t df, uint64_t tf) { scoreData.emplace_back(df, tf); }
+};
+
+struct FTSEdgeCompute : public EdgeCompute {
+    DoublePathLengthsFrontierPair* termsFrontier;
+    common::node_id_map_t<ScoreInfo>* scores;
+    common::node_id_map_t<uint64_t>* dfs;
+
+    FTSEdgeCompute(DoublePathLengthsFrontierPair* termsFrontier,
+        common::node_id_map_t<ScoreInfo>* scores, common::node_id_map_t<uint64_t>* dfs)
+        : termsFrontier{termsFrontier}, scores{scores}, dfs{dfs} {}
+
+    std::vector<nodeID_t> edgeCompute(nodeID_t boundNodeID, graph::NbrScanState::Chunk& resultChunk,
+        bool) override;
+
+    std::unique_ptr<EdgeCompute> copy() override {
+        return std::make_unique<FTSEdgeCompute>(termsFrontier, scores, dfs);
+    }
+};
+
+std::vector<nodeID_t> FTSEdgeCompute::edgeCompute(nodeID_t boundNodeID,
+    graph::NbrScanState::Chunk& resultChunk, bool) {
+    KU_ASSERT(dfs->contains(boundNodeID));
+    std::vector<nodeID_t> activeNodes;
+    resultChunk.forEach<uint64_t>([&](auto docNodeID, auto /* edgeID */, auto tf) {
+        auto df = dfs->at(boundNodeID);
+        if (!scores->contains(docNodeID)) {
+            scores->emplace(docNodeID, ScoreInfo{boundNodeID});
+        }
+        scores->at(docNodeID).addEdge(df, tf);
+        activeNodes.push_back(docNodeID);
+    });
+    return activeNodes;
+}
+
+struct FTSOutput {
+    common::node_id_map_t<ScoreInfo> scores;
+
+    FTSOutput() = default;
+    virtual ~FTSOutput() = default;
+};
+
+struct FTSState : public function::GDSComputeState {
+    void initTerms(processor::GDSCallSharedState& sharedState, RoaringBitmapSemiMask& termsMask,
+        transaction::Transaction* tx, common::table_id_t termsTableID) const;
+
+    FTSState(std::unique_ptr<function::FrontierPair> frontierPair,
+        std::unique_ptr<function::EdgeCompute> edgeCompute, common::table_id_t termsTableID);
+};
+
+FTSState::FTSState(std::unique_ptr<function::FrontierPair> frontierPair,
+    std::unique_ptr<function::EdgeCompute> edgeCompute, common::table_id_t termsTableID)
+    : function::GDSComputeState{std::move(frontierPair), std::move(edgeCompute)} {
+    this->frontierPair->getNextFrontierUnsafe().ptrCast<PathLengths>()->fixNextFrontierNodeTable(
+        termsTableID);
+}
+
+void FTSState::initTerms(processor::GDSCallSharedState& sharedState,
```

Comment:
I would call this initFirstFrontierWithTerms. This is initializing the frontier and not the terms.